src/lib.rs

```diff
@@ -1,41 +1,480 @@
+                item.info.distributor,
+                item.info.price,
+            )
+            .await;
+            ItemState::ForSaleByProducer
+        };
+
+        reply(SupplyChainEvent::Success);
+    }
+
+    fn ship_by_producer(&mut self, item_id: ItemId) {
+        self.check_producer();
+        let item = get_mut_item(&mut self.items, item_id);
+        assert_eq!(item.info.state, ItemState::ApprovedByProducer);
+        assert_eq!(item.info.producer, msg::source());
+
+        item.shipping_time = exec::block_timestamp();
+
+        item.info.state = ItemState::ShippedByProducer;
+        reply(SupplyChainEvent::Success);
+    }
+
+    async fn receive_by_distributor(&mut self, item_id: ItemId) {
+        self.check_distributor();
+        let item = get_mut_item(&mut self.items, item_id);
+        assert_eq!(item.info.state, ItemState::ShippedByProducer);
+        assert_eq!(item.info.distributor, msg::source());
 
-    use gtest::{Log, Program, System};
+        receive(self.ft_program_id, item.info.producer, item).await;
+        transfer_nft(self.nft_program_id, msg::source(), item_id).await;
 
-    #[test]
-    fn it_works() {
-        let system = System::new();
-        system.init_logger();
+        item.info.state = ItemState::ReceivedByDistributor;
+        reply(SupplyChainEvent::Success);
+    }
+
+    fn process_by_distributor(&mut self, item_id: ItemId) {
+        self.check_distributor();
+        let item = get_mut_item(&mut self.items, item_id);
+        assert_eq!(item.info.state, ItemState::ReceivedByDistributor);
+        assert_eq!(item.info.distributor, msg::source());
+
+        item.info.state = ItemState::ProcessedByDistributor;
+        reply(SupplyChainEvent::Success);
+    }
+
+    fn package_by_distributor(&mut self, item_id: ItemId) {
+        self.check_distributor();
+        let item = get_mut_item(&mut self.items, item_id);
+        assert_eq!(item.info.state, ItemState::ProcessedByDistributor);
+        assert_eq!(item.info.distributor, msg::source());
+
+        item.info.state = ItemState::PackagedByDistributor;
+        reply(SupplyChainEvent::Success);
+    }
+
+    async fn put_up_for_sale_by_distributor(&mut self, item_id: ItemId, price: u128) {
+        self.check_distributor();
+        let item = get_mut_item(&mut self.items, item_id);
+        assert_eq!(item.info.state, ItemState::PackagedByDistributor);
+        assert_eq!(item.info.distributor, msg::source());
+
+        item.info.price = price;
+        transfer_nft(self.nft_program_id, exec::program_id(), item_id).await;
+
+        item.info.state = ItemState::ForSaleByDistributor;
+        reply(SupplyChainEvent::Success);
+    }
+
+    async fn purchase_by_retailer(&mut self, item_id: ItemId, delivery_time: u64) {
+        self.check_retailer();
+        let item = get_mut_item(&mut self.items, item_id);
+        assert_eq!(item.info.state, ItemState::ForSaleByDistributor);
+
+        transfer_tokens(
```

Comment:
In my opinion it's better to send tokens when the sale is approved (the same applies to `purchase_by_distributor`, `purchase_by_consumer`)

Reply:
Sadly, due to the fact that the "approve" action in the FT contract approves transfers only for `msg::source()`, it's impossible to approve tokens transfers from a buyer to a supply chain on behalf of the seller inside the supply chain contract.

https://github.com/gear-dapps/fungible-token/blob/4da67648659e58dc2c60a9a589cdfc4f38d58329/src/lib.rs#L101

It's possible to do this externally, if the buyer himself approves transfers to the supply chain, but this will obviously be very inconvenient.